commit bf38e07f4dbc1d0511c1e0b4559f61673fa21b61 (refs/changes/76/1976/2)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-06-23T11:32:32-07:00 (2 years, 4 months ago)
    
    MANTA-3274 show how to run sdc-manta 'make test' when developing on non-smartos

diff --git a/.gitignore b/.gitignore
index 2e23092..47802ec 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,6 @@
 /node_modules
 /tmp
+/etc
 build
 docs/*.json
 docs/*.html
diff --git a/Makefile b/Makefile
index 461a7de..483a28e 100644
--- a/Makefile
+++ b/Makefile
@@ -59,7 +59,16 @@ include ./tools/mk/Makefile.node_deps.defs
 NODE_PREBUILT_VERSION=v0.10.32
 NODE_PREBUILT_TAG=zone
 NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
-include ./tools/mk/Makefile.node_prebuilt.defs
+ifeq ($(shell uname -s),SunOS)
+	include ./tools/mk/Makefile.node_prebuilt.defs
+	NODE_EXEC_DIR=$(TOP)/build/node/bin
+else
+	NPM=npm
+	NODE=node
+	NPM_EXEC=$(shell which npm)
+	NODE_EXEC=$(shell which node)
+	NODE_EXEC_DIR=$(shell dirname $(NODE_EXEC))
+endif
 
 MAN_INROOT	 = docs/man
 MAN_OUTROOT	 = man
@@ -105,7 +114,7 @@ prepush: check-probe-files
 
 .PHONY: test
 test: | $(CATEST)
-	PATH="$(TOP)/build/node/bin:$$PATH" $(CATEST) -a
+	PATH="$(NODE_EXEC_DIR):$$PATH" $(CATEST) -a
 
 $(CATEST): deps/catest/.git
 
@@ -155,7 +164,9 @@ publish: release
 CLEAN_FILES += node_modules
 
 include ./tools/mk/Makefile.deps
-include ./tools/mk/Makefile.node_prebuilt.targ
+ifeq ($(shell uname -s),SunOS)
+	include ./tools/mk/Makefile.node_prebuilt.targ
+endif
 include ./tools/mk/Makefile.node_deps.targ
 include ./tools/mk/Makefile.targ
 
diff --git a/README.md b/README.md
index 0b81d28..faa406c 100644
--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2016, Joyent, Inc.
+    Copyright (c) 2017, Joyent, Inc.
 -->
 
 # sdc-manta
@@ -59,7 +59,26 @@ and if warranted, get a code review.
 configuration file in etc/config.json.  The easiest way to create one is to copy
 the template in sapi\_manifests/manta/template and fill in the details for your
 environment, or else copy the configuration file directly out of a deployed
-sdc-manta zone in your environment.
+sdc-manta zone in your environment.  This assumes that you are developing on
+SmartOS. Read the next section if not.
+
+
+## `make prepush` if you build on non-SmartOS
+
+If you build on, say, MacOS, you will not be able to run the test suite
+(which is part of `make prepush`). An alternative is to:
+
+1. run `make prepush` on your Mac (ignoring the test failures); and
+
+2. sync your local changes to a deployed `manta0` zone (e.g. in COAL) and test
+   there. This can be done as follows:
+
+        ./tools/rsync-to $HEADNODE   # e.g. ./tools/rsync-to root@10.99.99.7
+        ssh $HEADNODE
+        sdc-login -l manta
+        cd /opt/smartdc/manta-deployment
+        pkgin in -y make
+        make test
 
 
 ## Adding a new Manta service
diff --git a/tools/rsync-to b/tools/rsync-to
new file mode 100755
index 0000000..382216f
--- /dev/null
+++ b/tools/rsync-to
@@ -0,0 +1,39 @@
+#!/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright (c) 2017, Joyent, Inc.
+#
+
+#
+# Rsync the master in this working copy to the install on the given HN.
+#
+
+#set -o xtrace
+set -o errexit
+
+TOP=$(cd $(dirname $0)/../; pwd)
+NODE=$1
+
+if [[ -z "$MANTA_ZONE" ]]; then
+    MANTA_ZONE=$(ssh $NODE "vmadm lookup -1 alias=manta0" 2>/dev/null)
+fi
+echo "MANTA_ZONE: $MANTA_ZONE"
+
+extraOpts=
+if [[ $(uname -s) != "SunOS" ]]; then
+    extraOpts="--exclude *.node --exclude build"
+fi
+
+# Note that we include tools/ and deps/ here to support running the test suite
+# in the rsync'd-to manta0 zone.
+rsync -av ${TOP}/ \
+    $NODE:/zones/$MANTA_ZONE/root/opt/smartdc/manta-deployment/ \
+    $extraOpts \
+    --exclude .git/ \
+    --exclude /boot/ \
+    --exclude /tmp/
